### PR TITLE
[Ready for Review] Track in-flight solving tx to delay HTLC failure update until enough confirmations

### PR DESCRIPTION
**Pull request by ariard**
WIP, need to implement block_disconnected, fix tests (was expected they break) and surely prune duplicate stuff in is_resolving_htlc_output.

### CODE REVIEW

**Comment by ariard** [comment]
Al casesl should be covered, just need to complete tests:

Delayed-Failures-until-maturation-cases :
* Revoked Remote HTLC: claim_htlc_outputs_shared_tx/claim_htlc_outpus_single_tx/test_simple_commitment_revoked_fail_backward/do_test_commitment_revoked_fail_backward_exhaustive
* Remote Dust HTLC: test_fail_backwards_latest_remote_announce/test_fail_backwards_previous_remote_announce
* Remote Non-Dust HTLC: test_fail_backwards_latest_remote_announce/test_fail_backwards_previous_remote_announce
 * Local Dust HTLC: (doing)
 * Local Non-Dust HTLC:
test_htlc_on_chain_timeout/test_duplicate_payment_hash_one_failure_one_success

Canceled-Failures :
* Revoked Remote HTLC:
* Remote Dust HTLC:
* Remote Non-Dust HTLC:
* Local Dust HTLC:
* Local Non-Dust HTLC:

**Comment by ariard** [comment]
Heey, should be good now. Modify a lot of tests due to introduction of delay in both ChannelMonitor and ChannelManager. Hope I don't change semantic of them while fixing

**Comment by ariard** [comment]
Rebased, and fix full_stack_target break

File: src/ln/channelmonitor.rs [hunk 1]
```diff
@@ -1174,16 +1184,20 @@ impl ChannelMonitor {
...
-				// on-chain claims, so we can do that at the same time.
 				macro_rules! check_htlc_fails {
 					($txid: expr, $commitment_tx: expr) => {
 						if let Some(ref outpoints) = self.remote_claimable_outpoints.get(&$txid) {
 							for &(ref htlc, ref source_option) in outpoints.iter() {
 								if let &Some(ref source) = source_option {
-									log_trace!(self, "Failing HTLC with payment_hash {} from {} remote commitment tx due to broadcast of revoked remote commitment transaction", log_bytes!(htlc.payment_hash.0), $commitment_tx);
-									htlc_updated.push(((**source).clone(), None, htlc.payment_hash.clone()));
+									log_trace!(self, "Failing HTLC with payment_hash {} from {} remote commitment tx due to broadcast of revoked remote commitment transaction, waiting confirmation until {} height", log_bytes!(htlc.payment_hash.0), $commitment_tx, height + HTLC_FAIL_ANTI_REORG_DELAY - 1);
+									match self.htlc_updated_waiting_threshold_conf.entry(height + HTLC_FAIL_ANTI_REORG_DELAY - 1) {
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Well, there L1192, its revoked HTLC outputs which are concerned, so we really care for maturation of revoked commitment transaction, their htlc_expiry may already been expired ? And in other places it's from broadcasting of failure-trigger tx we start timer, so don't get you on min(htlc_expiry - some_buffer) ? 

Timely failure of inbound HTLC may triggered by some per-tx-that-claims (like a HTLC-timeout) but IMO it's 2 different problems : waiting delay A for maturation of failure-trigger tx and waiting delay B before bumping timeout/claim tx. If so, shouldn't be hard to track txn_broadcast in check_spend_local_tx/check_spend_remote_tx, and at height timer expiration regenerate tx with higher fee, but seems to me it merits its own PR.

File: src/ln/channelmonitor.rs [hunk 2]
```diff
@@ -172,10 +172,6 @@ impl<Key : Send + cmp::Eq + hash::Hash> ChainListener for SimpleManyChannelMonit
 						// In case of reorg we may have htlc outputs solved in a different way so
 						// we prefer to keep claims but don't store duplicate updates for a given
 						// (payment_hash, HTLCSource) pair.
-						// TODO: Note that we currently don't really use this as ChannelManager
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/ln/channelmonitor.rs]: Okay, move the de-duplicating code at ChannelMonitor, a little bit simpler (even if not most efficient way due to HahsMap<HTLCSource, update> being non-implementable due to Hash trait absent on SecretKey)

**Comment by ariard** [comment]
Rebased, bf7f26a differs slightly from its previous state due to the fact I forget at first to prune channel closing in case of block disconnected

**Comment by ariard** [comment]
Closing, #336 is doing the job
